subsys/bluetooth/mesh/prov.c

```diff
@@ -107,10 +107,98 @@ static bt_mesh_input_action_t input_action(uint8_t action)
 	}
 }
 
-int bt_mesh_prov_auth(uint8_t method, uint8_t action, uint8_t size)
+static int check_output_auth(bt_mesh_output_action_t output, uint8_t size)
+{
+	if (!output) {
+		return -EINVAL;
+	}
+
+	if (!(bt_mesh_prov->output_actions & output)) {
+		return -EINVAL;
+	}
+
+	if (size > bt_mesh_prov->output_size) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int check_input_auth(bt_mesh_input_action_t input, uint8_t size)
+{
+	if (!input) {
+		return -EINVAL;
+	}
+
+	if (!(bt_mesh_prov->input_actions & input)) {
+		return -EINVAL;
+	}
+
+	if (size > bt_mesh_prov->input_size) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static void get_auth_string(unsigned char *str, uint8_t size)
+{
+	bt_rand(str, size);
+
+	/* Normalize to '0' .. '9' & 'A' .. 'Z' */
+	for (uint8_t i = 0U; i < size; i++) {
+		str[i] %= 36;
+		if (str[i] < 10) {
+			str[i] += '0';
+		} else {
+			str[i] += 'A' - 10;
+		}
+	}
+	str[size] = '\0';
+
+	memcpy(bt_mesh_prov_link.auth, str, size);
+	memset(bt_mesh_prov_link.auth + size, 0,
+			sizeof(bt_mesh_prov_link.auth) - size);
+}
+
+static uint32_t get_auth_number(bt_mesh_output_action_t output,
+		bt_mesh_input_action_t input, uint8_t size)
+{
+	uint32_t divider = 1;
+	uint32_t num = 0;
+
+	bt_rand(&num, sizeof(num));
+
+	for (uint8_t i = 1; i < size + 1; i++) {
+		divider *= 10;
+	}
+
+	if (output == BT_MESH_BLINK ||
+	    output == BT_MESH_BEEP ||
+	    output == BT_MESH_VIBRATE ||
+	    input == BT_MESH_PUSH ||
+	    input == BT_MESH_TWIST) {
+		/* According to the Bluetooth Mesh Profile
+		 * Specification Section 5.4.2.4, blink, beep
+		 * and vibrate should be a random integer
+		 * between 0 and 10^size, *exclusive*:
+		 */
+		num = (num % (divider - 1)) + 1;
+	} else {
+		num %= divider;
+	}
+
+	sys_put_be32(num, &bt_mesh_prov_link.auth[12]);
+	memset(bt_mesh_prov_link.auth, 0, 12);
+
+	return num;
+}
+
+int bt_mesh_prov_auth(bool is_provisioner, uint8_t method, uint8_t action, uint8_t size)
```

Comment:
This is really bad practice. This mimics what it is not really. Especially for output and input actions that are not abstract digit and string. Convert `BT_MESH_BLINK` to `BT_MESH_PUSH` and `BT_MESH_TWIST` to `BT_MESH_BEEP` etc etc. It has been implemented before and it was difficult to understand why `BT_MESH_BLINK` worked but `BT_MESH_PUSH` didn't and why the provisioner required size setting for method that device doesn't use. Less mimic magic.
Moving  `check_..._auth` outside prov.c causes one more huge `switch .. case` that will encumber provisioning device implementation with variety size checking. 
I'd prefer to leave what I refactored.

Reply:
I don't fully agree that it's bad practice, as long as we're clear on this function's purpose. If its purpose is "execute the authentication procedure for this provisioning session", I agree that it's odd to feed it the wrong values. If its purpose is "run this authentication procedure for me", as called by the provisioner and provisioning device, it makes perfect sense that the provisioner has its own logic to feed the algorithm, IMO. 

I favor the second interpretation, as the implementation in this patch is leaking role specific details into this module, which doesn't happen in other parts of this file.

Reply:
As I've already mentioned the authentication procedure is not fully symmetrical. There are more requirements to the provisioning device rather than to provisioner. It is not possible to use the same implementation as-is for both sides (btw. trying to use it caused all bugs that I fixed here). The number of authentication methods is not symmetrical depending on direction. When you mimic 3 output auth to 2 input auth it is not possible unequivocally to understand which output OOB is exactly used. Furthermore, the device requires more strict auth OOB sizes checking. Using "run this authentication procedure for me" approach requires exposing this functionality outside of prov.c
I'd agree with your design idea but it requires even more refactoring. Both sides provisioning device and provisioner should have its own switch case regarding OOB methods and call simple functions from prov.c like `generate number for blinking`, `generate general number`, `generate string`  etc.